### ISSUE TITLE
`OK` -> `OK` is being displayed as a hard state change

### ISSUE THREAD

**Issue by Yonas-net** [Icinga/icingadb#402]
## Describe the bug

Icinga DB shows `OK -> OK` as hard state change for random hosts in web as well as in DB. In the database `soft_state`, `hard_state` and previous_hard_state` have exactly the same value 0. And as `state_type` it states `hard`. Here screenshot of Icinga DB Web and monitoring module with the same host objects.

**Icinga DB Web:**
<img width="621" alt="Bildschirmfoto 2021-11-03 um 16 28 19" src="https://user-images.githubusercontent.com/57616252/140091115-3095ae8a-3f0f-4381-9211-c943b9c74e58.png">

**Monitoring:**
<img width="621" alt="Bildschirmfoto 2021-11-03 um 16 28 52" src="https://user-images.githubusercontent.com/57616252/140091211-5e2fffc3-a44f-4c5a-aa89-cd02cdf2ef98.png">

## To Reproduce

This always happens after I have Icinga 2 running for a couple of hours. It also doesn't happen with all host objects, but is just arbitrary.

**Comment by julianbrost** [comment]
So you would expect the the same history event but with `state_type=soft` instead of `hard`?

**Comment by Yonas-net** [comment]
> So you would expect the the same history event but with `state_type=soft` instead of `hard`?

Yes, and the three different states already mentioned above shouldn't actually have exactly the same value, either. At least `previous_hard_state` should never be the same as `hard_state` and `soft_state`.

**Comment by Yonas-net** [comment]
It's kind of strange, sometimes some objects don't get any status updates at all, despite the fact that in the IDO their status is updated all the time.

**Icinga DB:**
<img width="551" alt="Bildschirmfoto 2021-11-04 um 10 33 50" src="https://user-images.githubusercontent.com/57616252/140290933-dfff0812-e242-4088-89e9-c318d3670ffb.png"><img width="347" alt="Bildschirmfoto 2021-11-04 um 10 33 31" src="https://user-images.githubusercontent.com/57616252/140290999-590dec0b-a6ac-478d-b5cb-dc2f64325bdd.png">

**IDO:**
<img width="621" alt="Bildschirmfoto 2021-11-04 um 10 32 17" src="https://user-images.githubusercontent.com/57616252/140291065-eca363e2-1dde-46ec-b847-e5f8ddd069cd.png"> <img width="322" alt="Bildschirmfoto 2021-11-04 um 10 33 06" src="https://user-images.githubusercontent.com/57616252/140291078-f83674e1-61fc-4385-9606-8198a01879cf.png">

**Comment by julianbrost** [comment]
This happens due to the soft problem state before.

I think, at the moment, state history entries have to be interpreted like this: There was any state change and the new state is a `state_type` state, and not that the `state_type` state changed.

This also matches what's displayed in the monitoring module: the :+1: signals a hard state change.

So we get to the question whether this is fine as is or if we want to change it? I have a slight tendency towards keep it this way, as I think the result in IDB Web looks fine as is, but if someone can provide a good reason why it should be changed, I'd be open to that as well.

**Comment by lippserd** [comment]
Let's leave it as it is. DB Web could do something like `if hard state == 0 && previous hard state == 0: use previous soft state instead of previous hard state`. But previous soft state is missing for the state tables.

**Comment by julianbrost** [comment]
> DB Web could do something like `if hard state == 0 && previous hard state == 0: use previous soft state instead of previous hard state`.

I my following understanding of how IDB Web displays state history correct? The circle on the left/in the background represents the previous state, the one on the right/in the foreground the new state. And the new circle is smaller on soft states and full-sized on hard states.

So if that's correct, I think the ideal way to display this change would be a small red circle on the left and a big green circle on the right, i.e. also apply the soft/hard size to the one on the left. But this would probably need a new `previous_state_type` column or maybe the check you suggested could achieve the same result but somehow feels hacky.

> But previous soft state is missing for the state tables.

Huh? There is `state_history.previous_soft_state`?

**Comment by lippserd** [comment]
> Huh? There is `state_history.previous_soft_state`?

True. But it's missing in `host_state` and `service_state` which is used for the detail headers.
